File: frontend/web/src/types/discovery.ts
// TypeScript interfaces for Sprint 2.1 + 2.2 — Product Search & Discovery Layer

export interface ListingCandidate {
  id: string;
  marketplace: string;
  title: string;
  url: string;
  price: number;
  seller: string;
  thumbnail?: string | null;
  currency: string;
  availability: string;
  discovery_source: string;
  confidence: number;

  // ── Multi-Stage Discovery Confidence & Provenance (Refinements 3 & 4) ────────
  search_confidence?: number;
  matching_confidence?: number;
  discovery_confidence?: number;
  investigation_confidence?: number | null;
  discovered_via?: string;
  provenance_chain?: string[];

  metadata?: Record<string, unknown>;
}

export interface DiscoverySearchRequest {
  query: string;
  marketplaces?: string[] | null;
  limit_per_marketplace?: number;
  use_memory_cache?: boolean;
}

// ── Marketplace Health Score Interface ────────────────────────────────────────

export interface MarketplaceHealthInfo {
  marketplace: string;
  health_score: number;
  status: string;
  latency_ms: number;
  captcha_rate: number;
  data_quality_score: number;
}

// ── Sprint 2.2: Deduplication & Ranking types ─────────────────────────────

export interface PriorityScore {
  total_priority_score: number;        // 0–100
  price_anomaly_score: number;         // 0–1
  seller_trust_score: number;          // 0–1 (higher = less trusted)
  marketplace_risk_score: number;      // 0–1
  listing_completeness_score: number;  // 0–1 (higher = more incomplete)
  metadata_quality_score: number;      // 0–1 (higher = lower quality)
  reasoning: string[];
}

export type InvestigationPriority = 'critical' | 'high' | 'normal' | 'low';

export interface ListingGroup {
  group_id: string;
  canonical_title: string;
  normalized_product_name: string;
  listings: ListingCandidate[];
  representative?: ListingCandidate | null;
  unique_marketplaces: string[];
  unique_sellers: string[];
  price_range: { min?: number; max?: number; avg?: number };
  similarity_basis: string;
  listing_count: number;
  priority_score?: PriorityScore | null;
  investigation_priority: InvestigationPriority;
  created_at: string;
}

export interface DiscoverySearchResponse {
  query: string;
  normalized_query: string;
  marketplaces_searched: string[];
  // Sprint 2.1 — flat candidates list (preserved for backward compat)
  candidates: ListingCandidate[];
  // Sprint 2.2 — deduplicated + ranked groups
  listing_groups: ListingGroup[];
  top_investigation_targets: ListingCandidate[];
  metadata: {
    candidate_count: number;
    group_count?: number;
    top_target_count?: number;
    duration_ms: number;
    timestamp: string;
    search_engine_version: string;
    deduplication_reduction?: number;
    marketplace_health_scores?: Record<string, MarketplaceHealthInfo>;
    from_memory?: boolean;
    cached_at?: string;
  };
}

export interface SupportedMarketplacesResponse {
  supported_marketplaces: string[];
  count: number;
}

// ── Helper functions ──────────────────────────────────────────────────────────

export type ConfidenceTier = 'high' | 'medium' | 'low';

export function getConfidenceTier(confidence: number): ConfidenceTier {
  if (confidence >= 0.85) return 'high';
  if (confidence >= 0.65) return 'medium';
  return 'low';
}

export function formatConfidence(confidence: number): string {
  return `${Math.round(confidence * 100)}%`;
}

export function getPriorityColor(priority: InvestigationPriority): {
  bg: string; text: string; border: string;
} {
  switch (priority) {
    case 'critical': return { bg: 'bg-red-500/20', text: 'text-red-300', border: 'border-red-500/40' };
    case 'high':     return { bg: 'bg-orange-500/20', text: 'text-orange-300', border: 'border-orange-500/40' };
    case 'normal':   return { bg: 'bg-blue-500/20', text: 'text-blue-300', border: 'border-blue-500/40' };
    case 'low':      return { bg: 'bg-slate-500/20', text: 'text-slate-400', border: 'border-slate-600/40' };
  }
}

export function formatPriorityScore(score: number): string {
  return `${Math.round(score)}/100`;
}

// ── Sprint 2.3: Parallel Investigation Launcher types ─────────────────────────

export interface CandidateLaunchItem {
  candidate_id: string;
  marketplace: string;
  title: string;
  url: string;
  price?: number;
  seller?: string;
  currency?: string;
}

export interface ParallelLaunchRequest {
  candidates: CandidateLaunchItem[];
  investigation_type?: string;
  planner_strategy?: string;
  objectives?: string[];
  priority?: 'low' | 'medium' | 'high' | 'critical';
  notes?: string | null;
  advanced_options?: Record<string, unknown> | null;
}

export type JobStatus = 'pending' | 'in_progress' | 'completed' | 'failed' | 'unknown';

export interface LaunchJobStatus {
  candidate_id: string;
  investigation_id: string;
  marketplace: string;
  title: string;
  url: string;
  status: JobStatus;
  launched_at: string;
}

export interface ParallelLaunchResponse {
  batch_id: string;
  total_launched: number;
  jobs: LaunchJobStatus[];
  investigation_ids: string[];
  summary: string;
  metadata: {
    batch_id: string;
    launched_at: string;
    investigation_type: string;
    planner_strategy: string;
    candidate_count: number;
    marketplace_count: number;
  };
}

export interface BatchStatusResponse {
  batch_id: string;
  total: number;
  completed: number;
  in_progress: number;
  pending: number;
  failed: number;
  progress_pct: number;
  jobs: LaunchJobStatus[];
  is_complete: boolean;
}

// Convert a ListingCandidate to a CandidateLaunchItem
export function candidateToLaunchItem(c: ListingCandidate): CandidateLaunchItem {
  return {
    candidate_id: c.id,
    marketplace: c.marketplace,
    title: c.title,
    url: c.url,
    price: c.price,
    seller: c.seller,
    currency: c.currency,
  };
}

// ── Sprint 2.5: Product Intelligence Report types ─────────────────────────────

export interface ListingReportItem {
  investigation_id: string;
  marketplace: string;
  listing_url: string;
  title: string;
  seller: string;
  price: number;
  risk_score: number;
  verdict: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  confidence: number;
  evidence_count: number;
  top_risk_factor?: string | null;
  last_updated: string;
}

export interface ProductIntelligenceReportRequest {
  investigation_ids: string[];
  product_name?: string;
}

export interface ProductIntelligenceReport {
  report_id: string;
  product_name: string;
  generated_at: string;
  total_listings: number;
  safe_listings: number;
  suspicious_listings: number;
  overall_product_risk: number;
  overall_risk_level: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  highest_risk_marketplace: string;
  recommended_seller?: string | null;
  marketplace_distribution: Record<string, number>;
  evidence_summary: string[];
  coordinator_summary: string;
  investigations: ListingReportItem[];
  recommendations: string[];
  metadata?: Record<string, unknown>;
}

File: backend/schemas/discovery.py
import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class ListingCandidate(BaseModel):
    """
    Represents a discovered product candidate listing from a marketplace search adapter.
    Includes stage-specific confidence scores and full discovery provenance lineage.
    """

    id: str = Field(default_factory=lambda: f"cand-{uuid.uuid4().hex[:8]}")
    marketplace: str = "Global"
    title: str
    url: str
    price: float = 0.0
    seller: str = "Unverified Seller"
    thumbnail: Optional[str] = None
    currency: str = "INR"
    availability: str = "In Stock"
    discovery_source: str = "Marketplace Search Adapter"
    confidence: float = Field(0.85, ge=0.0, le=1.0)

    # ── Refinement 3: Multi-Stage Discovery Confidence ────────────────────────
    search_confidence: float = Field(
        0.90, ge=0.0, le=1.0, description="Query match quality score"
    )
    matching_confidence: float = Field(
        0.85, ge=0.0, le=1.0, description="Deduplication & entity match quality"
    )
    discovery_confidence: float = Field(
        0.88, ge=0.0, le=1.0, description="Overall discovery confidence"
    )
    investigation_confidence: Optional[float] = Field(
        None, ge=0.0, le=1.0, description="Verdict confidence when investigated"
    )

    # ── Refinement 4: Discovery Provenance Lineage ────────────────────────────
    discovered_via: str = Field(
        "Search",
        description="Marketplace API | Search | Scraper | Manual | Historical Memory",
    )
    provenance_chain: List[str] = Field(
        default_factory=lambda: ["Marketplace API: Search Adapter"],
        description="Lineage trace steps explaining discovery origin",
    )

    metadata: Dict[str, Any] = Field(default_factory=dict)

    # ── Feature 4: Full Candidate Lineage Metadata ─────────────────────────────
    candidate_id: Optional[str] = None
    http_request_id: Optional[str] = None
    response_sha256: Optional[str] = None
    evidence_archive_id: Optional[str] = None
    parser_version: str = Field("v1.2.0-bs4", description="Parser engine version")
    parser_confidence: float = Field(
        95.0, description="Parser extraction confidence score (0-100)"
    )
    retrieval_mode: str = Field(
        "LIVE_HTTP",
        description="LIVE_HTTP | OFFICIAL_API | CACHE | HISTORICAL_MEMORY | FALLBACK",
    )
    deduplication_group_id: Optional[str] = None
    ranking_score: float = 0.0
    investigation_id: Optional[str] = None
    report_id: Optional[str] = None


class DiscoverySearchRequest(BaseModel):
    query: str
    marketplaces: Optional[List[str]] = None
    limit_per_marketplace: int = Field(5, ge=1, le=20)
    use_memory_cache: bool = Field(
        True, description="Load previous discovery memory if available"
    )


# ── Refinement 2: Marketplace Health Score Schema ────────────────────────────


class MarketplaceHealthInfo(BaseModel):
    marketplace: str
    health_score: int = Field(
        ..., ge=0, le=100, description="Live reliability / health score 0-100"
    )
    status: str = "Operational"  # "Operational" | "Degraded" | "Offline"
    latency_ms: float = 120.0
    captcha_rate: float = 0.0
    data_quality_score: int = 95


# ── Sprint 2.2: Deduplication & Ranking Models ────────────────────────────────


class PriorityScore(BaseModel):
    """
    Composite priority score explaining why a listing group is ranked high for investigation.
    All component scores are in [0, 1] range; total_priority_score is in [0, 100].
    """

    total_priority_score: float = Field(0.0, ge=0.0, le=100.0)
    # Component sub-scores
    price_anomaly_score: float = Field(0.0, ge=0.0, le=1.0)
    seller_trust_score: float = Field(0.0, ge=0.0, le=1.0)  # higher = less trusted
    marketplace_risk_score: float = Field(0.0, ge=0.0, le=1.0)
    listing_completeness_score: float = Field(
        0.0, ge=0.0, le=1.0
    )  # higher = more incomplete
    metadata_quality_score: float = Field(0.0, ge=0.0, le=1.0)  # higher = lower quality
    # Human-readable breakdown
    reasoning: List[str] = Field(default_factory=list)


class ListingGroup(BaseModel):
    """
    A cluster of deduplicated equivalent listings for the same underlying product variant.
    Produced by the DeduplicationService and enriched by the RankingEngine.
    """

    group_id: str = Field(default_factory=lambda: f"grp-{uuid.uuid4().hex[:8]}")
    canonical_title: str
    normalized_product_name: str
    # All deduplicated candidates in this group (across marketplaces)
    listings: List[ListingCandidate]
    # Best representative listing for the group (highest confidence)
    representative: Optional[ListingCandidate] = None
    # Cross-listing deduplication metadata
    unique_marketplaces: List[str] = Field(default_factory=list)
    unique_sellers: List[str] = Field(default_factory=list)
    price_range: Dict[str, float] = Field(
        default_factory=dict
    )  # {"min": x, "max": y, "avg": z}
    similarity_basis: str = (
        "title"  # "title" | "model_number" | "seller" | "specification"
    )
    listing_count: int = 0
    # Ranking output
    priority_score: Optional[PriorityScore] = None
    investigation_priority: str = "normal"  # "critical" | "high" | "normal" | "low"
    created_at: str = Field(default_factory=lambda: datetime.utcnow().isoformat())


class DiscoverySearchResponse(BaseModel):
    query: str
    normalized_query: str
    marketplaces_searched: List[str]
    # Raw flat candidate list (Sprint 2.1 — preserved for backward compat)
    candidates: List[ListingCandidate]
    # Sprint 2.2: Deduplicated groups ranked by investigation priority
    listing_groups: List[ListingGroup] = Field(default_factory=list)
    # Top N candidates ranked for immediate investigation
    top_investigation_targets: List[ListingCandidate] = Field(default_factory=list)
    metadata: Dict[str, Any] = Field(default_factory=dict)
